---
title: "Microhaplotype analysis"
author: "Jorge Eduardo Amaya Romero"
date: "2022-09-27"
output:
  html_document:
    toc: yes
    number_sections: yes
    toc_float: yes
    code_folding: show
editor_options:
  markdown:
    wrap: 72
  chunk_output_type: inline
params:
  cigar_paths: "placeholder"
  metadata_source: "placeholder"
  ls_locus_remove: "placeholder"
  sample_id_pattern: "placeholder"
  pop_colors: "placeholder"
  pop_levels: "placeholder"
  path_to_markers: "placeholder"
  gene_names: "placeholder"
  gff_file: "placeholder"
  reference_alleles: "placeholder"
  fasta_file: "placeholder"
  gene_ids: "placeholder"
  variables: "placeholder"
  collection_quarter: "placeholder"
---

The following report provides a summary of analysis of genotyping information generated through amplicon sequencing (also called Microhaplotype genotyping).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r 'Loading libraries and parameters'}
source('/amplseq_required_libraries.R')
source('/amplseq_functions.R')

cigar_paths <- params$cigar_paths
metadata_source <- params$metadata_source
ls_locus_remove <- params$ls_locus_remove
sample_id_pattern <- params$sample_id_pattern
pop_colors <- unlist(strsplit(params$pop_colors, ","))
pop_levels <- adjust_camel_case_name(unlist(strsplit(params$pop_levels, ",")))
pop_levels = factor(pop_levels, levels = pop_levels)
path_to_markers <- params$path_to_markers

if ("pop_pairs" %in% names(params)) {
 pop_pairs <- params$pop_pairs
} else {
 pop_pairs = combn(sort(as.character(pop_levels)), 2, function(x) paste(x, collapse = "-"))
}
gene_names <- params$gene_names
gff_file <- params$gff_file
reference_alleles <- params$reference_alleles
fasta_file <- params$fasta_file
gene_ids <- params$gene_ids
variables <- params$variables
collection_quarter <- params$collection_quarter
filters = c(paste("Subnational_level2;", paste(pop_levels, collapse = ",")),
            paste("Quarter_of_Collection;", paste(collection_quarter, collapse = ",")))
```

```{r}
#Load the cigar objects and metadata and merge them
cigar_object = read_cigar_tables(paths = cigar_paths, sample_id_pattern = sample_id_pattern)
metadata = read.csv(metadata_source)
print(metadata_source)

cigar_object@metadata = merge(cigar_object@metadata,
                               metadata,
                               by = 'Sample_id',
                               all.x = T)

#Remove certain locus if requested
if (file.exists(ls_locus_remove)) {
  loci <- read.table(ls_locus_remove, header = FALSE, colClasses = "character", stringsAsFactors = FALSE)$V1
  for (locus in loci) {
    cigar_object@cigar_table = cigar_object@cigar_table[,!grepl(locus, colnames(cigar_object@cigar_table))]
  }
}

# Define the geographic unit of analysis and assign them colors
if (length(pop_colors) == length(pop_levels)) {
  pop_colors = pop_colors
} else {
  pop_colors = replicate(n, paste0("#", paste0(sample(0:255, 3, replace = TRUE), collapse = "")))
}
names(pop_levels) = pop_colors

print(pop_levels)
```

# Genotyped samples by Municipality over time
```{r}
# Plot the sample size (number of genotyped samples) by Municipality over time.
# Quarterly plot
dates = sort(unique(cigar_object@metadata$Quarter_of_Collection))
print(dates)

plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Quarter_of_Collection)) %>%
  ggplot(aes(x = Quarter_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                    levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Quarter") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_quarterly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()

```

The following plots show the sample size (number of genotyped samples) by Municipality over Quarters of Collection
```{r}
plot_temporal_collection_of_samples
```

```{r}
dates = sort(unique(cigar_object@metadata$Month_of_Collection))
plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Month_of_Collection)) %>%
  ggplot(aes(x = Month_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                    levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Month") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_monthly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()
```

The following plots show the sample size (number of genotyped samples) by Municipality over Months of Collection
```{r}
plot_temporal_collection_of_samples
```

```{r}
# Yearly plot
dates = sort(unique(cigar_object@metadata$Year_of_Collection))
plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Year_of_Collection)) %>%
  ggplot(aes(x = Year_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                  levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Year") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_yearly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()
```

The following plots show the sample size (number of genotyped samples) by Municipality over Year of Collection
```{r}
plot_temporal_collection_of_samples
```

# Amplification rate per loci

The proportion of samples that have been amplified by each amplicon marker. The threshold defines the minimum proportion of samples that a marker should amplify in order to be keep it for further analysis (default 0.65) ADD AN ARGUMENTS FOR THIS. All markers below this threshold will be discarded.

```{r}
markers = read.csv(path_to_markers)
ampseq = cigar2ampseq(cigar_object, markers = markers, min_abd = 10, min_ratio = .1, remove_controls = T)

ampseq_filtered = locus_amplification_rate(ampseq, threshold = .65)

# Plot Amplification rate per loci ----
plot_locus_amplificatin_rate = ggdraw()+
  draw_plot(ampseq_filtered@plots$amplification_rate_per_locus+
              theme(axis.text = element_text(size = 12),
                    axis.title = element_text(size = 12),
                    legend.text = element_text(size = 12)),
            x = 0,
            y = 0,
            width = 1,
            height = .5)+
  draw_plot(ampseq_filtered@plots$all_loci_amplification_rate+
              theme(axis.text = element_text(size = 12),
                    axis.title = element_text(size = 12)),
            x = 0,
            y = .5,
            width = 1,
            height = .5)

pdf("/cromwell_root/Results/plot_locus_amplificatin_rate.pdf")
plot_locus_amplificatin_rate
dev.off()
```

```{r}
plot_locus_amplificatin_rate
```

Thus in this data set `r ncol(ampseq@gt)` loci had an amplification rate above 0.65, and `r ncol(ampseq_filtered@discarded_loci$gt)` loci were discarded. The discarded loci were: `r paste(colnames(ampseq@discarded_loci$gt), collapse = ', ')`.

# Amplification rate by sample

The proportion of loci amplified per each sample, also called the amplification rate of the samples. The threshold defines the minimum proportion of loci that a sample should have amplified (by default 0.8). FIX THIS TO MAKE IT VARIABLE

```{r}
ampseq_filtered = sample_amplification_rate(ampseq_filtered)

pdf("/cromwell_root/Results/samples_amplification_rate.pdf")
ampseq_filtered@plots$samples_amplification_rate+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12))
dev.off()
```

```{r}
ampseq_filtered@plots$samples_amplification_rate+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12))
```

# plot_log_NPolyLoci_by_pop

WRITE THE TEXT FOR THIS PLOT

```{r}
ampseq_filtered = get_polygenomic(ampseq_object = ampseq_filtered, strata = "Subnational_level2", na.rm = FALSE,
                                  filters = NULL)

plot_log_NPolyLoci_by_pop = log_scale_histogram(data = ampseq_filtered@metadata[ampseq_filtered@metadata$NPolyLoci != 0,],
                                                var = "NPolyLoci", binwidth = 1, group_by = "Subnational_level2",
                                                levels = pop_levels, x_label = "Number of heterozygous loci per sample",
                                                fill_color = pop_colors,
                                                y_breaks = c(1, 5,10, 30), ncol = 4)+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12),
        strip.text = element_text(size = 12)) +
  labs(title = 'Distribution of number of heterozygous loci per sample',
       y = "# Samples",
       x = "# heterozygous loci per sample")

pdf("/cromwell_root/Results/distribution_of_number_heterozygous_loci_per_sample.pdf")
plot_log_NPolyLoci_by_pop
dev.off()
```

```{r}
plot_log_NPolyLoci_by_pop
```

#plot_log_coi_by_pop
WRITE THE TEXT FOR THIS PLOT

```{r}
plot_log_coi_by_pop = log_scale_histogram(data = ampseq_filtered@metadata,
                                          var = "coi",
                                          binwidth = 1,
                                          group_by = "Subnational_level2",
                                          levels = pop_levels, x_label = "COI",
                                          fill_color = pop_colors,
                                          y_breaks = c(1, 10, 100, 400), ncol = 5)+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12),
        strip.text = element_text(size = 12)) +
  labs(title = 'Distribution of COI by sampling location',
       y = "# Samples")

pdf("/cromwell_root/Results/distribution_of_COI_by_sampling_location.pdf")
plot_log_coi_by_pop
dev.off()
```

```{r}
plot_log_coi_by_pop
```

#plot_poly_by_pop_over_time

```{r}
ampseq_filtered@metadata %<>% mutate(Pop_quarter = paste(Subnational_level2, Quarter_of_Collection, sep = '_'))

plot_poly_by_pop_over_time = get_polygenomic(ampseq_object = ampseq_filtered, strata = "Pop_quarter", update_popsummary = F, na.rm = TRUE, filters = NULL)

plot_poly_by_pop_over_time = plot_poly_by_pop_over_time %>% filter(grepl(paste(pop_levels, collapse = "|"), pop))%>%
  mutate(
    Population = stringr::str_split(pop, '_', simplify = TRUE)[,1],
    Date = stringr::str_split(pop, '_', simplify = TRUE)[,2],
    prop_poly_lower = case_when(
      prop_poly == 0 ~ 0,
      prop_poly != 0 ~ prop_poly_lower),
    prop_poly_upper = case_when(
      prop_poly == 0 ~ 0,
      prop_poly != 0 ~ prop_poly_upper)
  )%>%
  ggplot(aes(x = Date,
             y = prop_poly,
             ymin = prop_poly_lower,
             ymax = prop_poly_upper,
             fill = factor(Population, levels = pop_levels)))+
  geom_col()+
  geom_errorbar(width = .2)+
  facet_wrap(~factor(Population, levels = pop_levels), ncol = 5)+
  theme_bw()+
  scale_fill_manual(values = pop_colors)+
  labs(title = 'Temporal change of the proportion of polyclonal infections',
       y = "Polyclonal infections",
       x = "Date of Collection")+
  theme(axis.text = element_text(size = 10),
        axis.text.x = element_text(angle = 315, vjust = 0),
        axis.title = element_text(size = 12),
        strip.text = element_text(size = 12),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_poly_by_pop_over_time.pdf")
plot_poly_by_pop_over_time
dev.off()
```

```{r}
plot_poly_by_pop_over_time
```

# Tested up to this point

# Distribution of relatedness by study area
FIX THE TABLE SHOWING. ASK PAULO.
RETHINK THE IMPLIMENTIATION OF THE PAIRWISE RELATEDNESS ALGORITHM BECAUSE OF CALCULATION TIMES

```{r}
# generating the loci_object----
loci_object = ampseq2loci(ampseq_filtered)

# measuring relatedness----
pairwise_relatedness = NULL

pairwise_relatedness = read.csv('/cromwell_root/mhap_metadata/col_pairwise_relatedness.csv')

#for(w in 1:20){
#  start = Sys.time()
#  pairwise_relatedness = rbind(pairwise_relatedness,
#                               pairwise_hmmIBD(loci_object, parallel = FALSE, w = w, n = 500))
#  time_diff = Sys.time() - start
#  print(paste0('step ', w, ' done in ', time_diff, ' secs'))
#}
write.csv(pairwise_relatedness, 'Results/col_pairwise_relatedness.csv', row.names = F, quote = F)

plot_relatedness_distribution_within = plot_relatedness_distribution(
  pairwise_relatedness = pairwise_relatedness,
  metadata = ampseq_filtered@metadata,
  Population = 'Subnational_level2',
  fill_color = pop_colors,
  type_pop_comparison = 'within',
  ncol = 5,
  pop_levels = pop_levels
)

pdf("/cromwell_root/Results/plot_relatedness_distribution_within.pdf", width = 14)
plot_relatedness_distribution_within$plot +
  labs(title = 'Distribution of relatedness by study area')
dev.off()
```

```{r}
plot_relatedness_distribution_within
```

# plot_frac_highly_related_within
```{r}
plot_frac_highly_related_within = plot_frac_highly_related(
  pairwise_relatedness = pairwise_relatedness,
  metadata = ampseq@metadata,
  Population = 'Subnational_level2',
  fill_color = pop_colors,
  threshold = 0.99, type_pop_comparison = 'within', pop_levels = pop_levels)

pdf("/cromwell_root/Results/plot_frac_highly_related_within.pdf")
plot_frac_highly_related_within$plot+
  labs(title = 'Highly related')
dev.off()
```

```{r}
plot_frac_highly_related_within$plot+
  labs(title = 'Highly related')
```

# Proportion of highly related samples over time (IBD > 0.99)
```{r}
plot_frac_highly_related_over_quarters_within = plot_frac_highly_related_over_time(pairwise_relatedness = pairwise_relatedness,
                                                                                   metadata = ampseq@metadata,
                                                                                   Population = c('Subnational_level2', 'Month_of_Collection'),
                                                                                   fill_color = pop_colors,
                                                                                   threshold = 0.99,
                                                                                   type_pop_comparison = 'within',
                                                                                   pop_levels = pop_levels, ncol = 1)

pdf("/cromwell_root/Results/plot_frac_highly_related_over_quarters_within.pdf")
plot_frac_highly_related_over_quarters_within$plot_frac_highly_related +
  labs(title = 'Proportion of highly related samples over time (IBD > 0.99)',
       y = 'Proportion')
dev.off()
```

```{r}
plot_frac_highly_related_over_quarters_within$plot_frac_highly_related +
  labs(title = 'Proportion of highly related samples over time (IBD > 0.99)',
       y = 'Proportion')
```

# plot_cor_highlyR_nsamples
```{r}
nsamples_vs_highlyR = left_join(plot_frac_highly_related_over_quarters_within$plot_frac_highly_related$data %>% mutate(pop = paste(Pop_comparison, Date_Yi, sep = '_'))%>% ungroup()%>%
                                  select(pop, prop),
                                  metadata %>%
                                  filter(!is.na(Subnational_level2)) %>%
                                  group_by(Subnational_level2, Month_of_Collection) %>%
                                  summarise(nsamples= n()) %>% mutate(pop = paste(Subnational_level2, Month_of_Collection, sep = '_')), by = 'pop')  %>%
                                  filter(prop != 0, Subnational_level2 %in% c(pop_levels))

plot_cor_highlyR_nsamples = ggscatter(nsamples_vs_highlyR, x = "prop", y = "nsamples",
                                      add = "reg.line", conf.int = TRUE,
                                      cor.coef = TRUE, cor.method = "spearman",
                                      title= 'Number of collected samples vs. Proportion of highly related samples',
                                      xlab = "Proportion of highly related samples", ylab = "Number of samples",
                                      facet.by = 'Subnational_level2')

pdf("/cromwell_root/Results/plot_cor_highlyR_nsamples.pdf")
plot_cor_highlyR_nsamples
dev.off()
```

```{r}
plot_cor_highlyR_nsamples
```

# plot_cor_proppoly_nsamples
```{r}
nsamples_vs_proppoly = left_join(plot_poly_by_pop_over_time$data %>% select(pop, prop_poly),
                                 metadata %>%
                                   mutate(Quarter_of_Collection = paste(substr(Date_of_Collection, 1, 4), quarters.Date(Date_of_Collection), sep='-'))%>%
                                   group_by(Subnational_level2, Quarter_of_Collection) %>%
                                   summarise(nsamples= n()) %>% mutate(pop = paste(Subnational_level2, Quarter_of_Collection, sep = '_')), by = 'pop') %>%
  filter(Subnational_level2 %in% c(pop_levels))

plot_cor_proppoly_nsamples = ggscatter(nsamples_vs_proppoly, x = "prop_poly", y = "nsamples",
                                       add = "reg.line", conf.int = TRUE,
                                       cor.coef = TRUE, cor.method = "spearman",
                                       title = 'Samples vs. Prop. Poly',
                                       xlab = "Proportion of polyclonal samples", ylab = "Number of samples", facet.by = 'Subnational_level2')

pdf("/cromwell_root/Results/plot_cor_proppoly_nsamples.pdf")
plot_cor_proppoly_nsamples
dev.off()
```

```{r}
plot_cor_proppoly_nsamples
```

# plot_network
```{r}
plot_network = plot_network(pairwise_relatedness = pairwise_relatedness,
                            threshold = .99,
                            metadata = ampseq_filtered@metadata,
                            sample_id = 'Sample_id',
                            group_by = 'Subnational_level2',
                            levels = pop_levels,
                            colors = pop_colors)

pdf("/cromwell_root/Results/plot_network.pdf")
plot_network
dev.off()
```

```{r}
plot_network
```

# plot_relatedness_distribution_between
FIX THE TABLE. ASK PAULO.
```{r}
plot_relatedness_distribution_between = plot_relatedness_distribution(
  pairwise_relatedness = pairwise_relatedness,
  metadata = ampseq_filtered@metadata,
  Population = 'Subnational_level2',
  fill_color = rep('gray50', 10),
  type_pop_comparison = 'between',
  ncol = 5,
  pop_levels = pop_pairs
)

pdf("/cromwell_root/Results/plot_relatedness_distribution_between.pdf", width = 14)
plot_relatedness_distribution_between$plot+
  labs(title = 'Distribution of relatedness')
dev.off()
```

```{r, echo=FALSE}
plot_relatedness_distribution_between$plot+
  labs(title = 'Distribution of relatedness')
```

# MAP
```{r}
## Polygon dot plot of observed G6PDd frequency by LAC country----
# Define the temporal unit of analysis
dates = sort(unique(cigar_object@metadata$Month_of_Collection))

plot_temporal_collection_of_samples = cigar_object@metadata %>%
 filter(!is.na(Subnational_level2)) %>%
 group_by(Subnational_level2, Month_of_Collection) %>%
 summarise(nsamples= n())%>%
 ggplot(aes(x = Month_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                 levels = pop_levels)))+
 geom_col()+
 theme_bw()+
 scale_fill_manual(values = pop_colors)+
 facet_wrap(.~factor(Subnational_level2,
                     levels = pop_levels),
            strip.position = "top", ncol = 1)+
 labs(title = 'Samples collected over time',
      y = "Collected samples by PCD",
      x = "Months")+
 theme(axis.text = element_text(size = 12),
       axis.text.x = element_text(angle = 315, vjust = 0),
       axis.title = element_text(size = 12),
       strip.text = element_text(size = 12),
       legend.position =  "none")+
 scale_alpha_manual(values = c(1,.6))+
 scale_x_discrete(limits = dates)
 
# Get all shape files from LAC countries at national level
# SpatialPolygonsDataFrame at country level (level = 0)
col0.spldf = gadm_sp_loadCountries("COL",level = 0, basefile = "./GADMTools/world.rds.files/rds0/")

# SpatialPolygonsDataFrame at Municipality level (level = 2)
col2.spldf = gadm_sp_loadCountries("COL",level = 2, basefile = "./GADMTools/world.rds.files/rds2/")

# Label for Municipalities
localities = data.frame(Localities = factor(pop_levels,
                                            levels = pop_levels[5:1]),
                        long = -70,
                        lat = c(6, 8, 10, 12, 14),
                        colors = pop_colors[5:1])

if ('Guapi' %in% pop_levels) {
  new_pop_levels = levels(pop_levels)
  new_pop_levels[which('Guapi' == new_pop_levels)] = 'Guapí'
  pop_levels = new_pop_levels
  pop_levels = factor(pop_levels, levels = pop_levels)
  names(pop_levels) = pop_colors
}

plot_study_sites = ggplot() +
  geom_polygon(data=col0.spldf$spdf, aes(x=long, y=lat, group = group), fill = "floralwhite", color = "gray40", linewidth=.3)

for (i in 1:length(pop_levels)) {
  print(pop_levels[i])
  print(names(pop_levels[i]))
  temp_spdf = col2.spldf[["spdf"]][col2.spldf[["spdf"]][["NAME_2"]] == pop_levels[i],]
  temp_polygon = data.frame(long = temp_spdf@polygons[[1]]@Polygons[[1]]@coords[,1],
                            lat = temp_spdf@polygons[[1]]@Polygons[[1]]@coords[,2])
  plot_study_sites = plot_study_sites +
    geom_polygon(data=col2.spldf[["spdf"]][col2.spldf[["spdf"]][["NAME_2"]] == pop_levels[i],], aes(x=long, y=lat, group = group), fill = names(pop_levels[i]), color = "gray40", linewidth=.3)
}

plot_study_sites = plot_study_sites +
  geom_point(data=localities, aes(x=0, y=90, color=Localities), size = 5) +
  theme_bw()+
  labs(title = "Study sites in Colombia",
       x = "Longitude",
       y = "Latitude")+
  scale_x_continuous(limits = c(-80, -65))+
  scale_y_continuous(limits = c(-5, 15))+
  scale_color_manual(values = localities$colors)+
  theme(legend.position = "right",
        text = element_text(size = 10),
        axis.text = element_text(size = 10),
        title = element_text(size = 10))+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12))
 
#Plot study areas
pdf("/cromwell_root/Results/plot_study_areas.pdf")
plot_study_sites
dev.off()
```

```{r}
plot_study_sites
```

Watermark 4
