---
title: "Microhaplotype analysis"
author: "Jorge Eduardo Amaya Romero"
date: "2022-09-27"
output:
  html_document:
    toc: yes
    number_sections: yes
    toc_float: yes
    code_folding: show
editor_options:
  markdown:
    wrap: 72
  chunk_output_type: inline
params:
  cigar_paths: "placeholder"
  metadata_source: "placeholder"
  ls_locus_remove: "placeholder"
  sample_id_pattern: "placeholder"
  pop_colors: "placeholder"
  pop_levels: "placeholder"
  path_to_markers: "placeholder"
  gene_names: "placeholder"
  gff_file: "placeholder"
  reference_alleles: "placeholder"
  fasta_file: "placeholder"
  gene_ids: "placeholder"
  variables: "placeholder"
  collection_quarter: "placeholder"
---

The following report provides a summary of analysis of genotyping information generated through amplicon sequencing (also called Microhaplotype genotyping).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r 'Loading libraries and parameters'}
source('/amplseq_required_libraries.R')
source('/amplseq_functions.R')

cigar_paths <- params$cigar_paths
metadata_source <- params$metadata_source
ls_locus_remove <- params$ls_locus_remove
sample_id_pattern <- params$sample_id_pattern
pop_colors <- unlist(strsplit(params$pop_colors, ","))
pop_levels <- adjust_camel_case_name(unlist(strsplit(params$pop_levels, ",")))
pop_levels = factor(pop_levels, levels = pop_levels)
path_to_markers <- params$path_to_markers

if ("pop_pairs" %in% names(params)) {
 pop_pairs <- params$pop_pairs
} else {
 pop_pairs = combn(sort(as.character(pop_levels)), 2, function(x) paste(x, collapse = "-"))
}
gene_names <- params$gene_names
gff_file <- params$gff_file
reference_alleles <- params$reference_alleles
fasta_file <- params$fasta_file
gene_ids <- params$gene_ids
variables <- params$variables
collection_quarter <- params$collection_quarter
```

```{r}
#Load the cigar objects and metadata and merge them
cigar_object = read_cigar_tables(paths = cigar_paths, sample_id_pattern = sample_id_pattern)
metadata = read.csv(metadata_source)
print(metadata_source)

cigar_object@metadata = merge(cigar_object@metadata,
                               metadata,
                               by = 'Sample_id',
                               all.x = T)

#Remove certain locus if requested
if (file.exists(ls_locus_remove)) {
  loci <- read.table(ls_locus_remove, header = FALSE, colClasses = "character", stringsAsFactors = FALSE)$V1
  for (locus in loci) {
    cigar_object@cigar_table = cigar_object@cigar_table[,!grepl(locus, colnames(cigar_object@cigar_table))]
  }
}

# Define the geographic unit of analysis and assign them colors
if (length(pop_colors) == length(pop_levels)) {
  pop_colors = pop_colors
} else {
  pop_colors = replicate(n, paste0("#", paste0(sample(0:255, 3, replace = TRUE), collapse = "")))
}
names(pop_levels) = pop_colors

print(pop_levels)
```

# Genotyped samples by Municipality over time
```{r}
# Plot the sample size (number of genotyped samples) by Municipality over time.
# Quarterly plot
dates = sort(unique(cigar_object@metadata$Quarter_of_Collection))
print(dates)

plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Quarter_of_Collection)) %>%
  ggplot(aes(x = Quarter_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                    levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Quarter") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_quarterly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()

```

The following plots show the sample size (number of genotyped samples) by Municipality over Quarters of Collection
```{r}
plot_temporal_collection_of_samples
```

```{r}
dates = sort(unique(cigar_object@metadata$Month_of_Collection))
plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Month_of_Collection)) %>%
  ggplot(aes(x = Month_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                    levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Month") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_monthly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()
```

The following plots show the sample size (number of genotyped samples) by Municipality over Months of Collection
```{r}
plot_temporal_collection_of_samples
```

```{r}
# Yearly plot
dates = sort(unique(cigar_object@metadata$Year_of_Collection))
plot_temporal_collection_of_samples = cigar_object@metadata %>%
  filter(!is.na(Subnational_level2)) %>% # Remove samples with no geographic information (Controls)
  summarise(nsamples= n(), .by = c(Subnational_level2, Year_of_Collection)) %>%
  ggplot(aes(x = Year_of_Collection, y = nsamples, fill = factor(Subnational_level2,
                                                                  levels = pop_levels))) +
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = pop_colors) +
  facet_wrap(.~factor(Subnational_level2,
                      levels = pop_levels),
             strip.position = "top", ncol = 5) +
  labs(title = 'Number of Samples collected over time',
       y = "Collected samples by PCD",
       x = "Year") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position =  "none")

pdf("/cromwell_root/Results/plot_temporal_collection_of_samples_yearly.pdf", width = 14)
plot_temporal_collection_of_samples
dev.off()
```

The following plots show the sample size (number of genotyped samples) by Municipality over Year of Collection
```{r}
plot_temporal_collection_of_samples
```

# Amplification rate per loci

The proportion of samples that have been amplified by each amplicon marker. The threshold defines the minimum proportion of samples that a marker should amplify in order to be keep it for further analysis (default 0.65) ADD AN ARGUMENTS FOR THIS. All markers below this threshold will be discarded.

```{r}
markers = read.csv(path_to_markers)
ampseq = cigar2ampseq(cigar_object, markers = markers, min_abd = 10, min_ratio = .1, remove_controls = T)

ampseq_filtered = locus_amplification_rate(ampseq, threshold = .65)

# Plot Amplification rate per loci ----
plot_locus_amplificatin_rate = ggdraw()+
  draw_plot(ampseq_filtered@plots$amplification_rate_per_locus+
              theme(axis.text = element_text(size = 12),
                    axis.title = element_text(size = 12),
                    legend.text = element_text(size = 12)),
            x = 0,
            y = 0,
            width = 1,
            height = .5)+
  draw_plot(ampseq_filtered@plots$all_loci_amplification_rate+
              theme(axis.text = element_text(size = 12),
                    axis.title = element_text(size = 12)),
            x = 0,
            y = .5,
            width = 1,
            height = .5)

pdf("/cromwell_root/Results/plot_locus_amplificatin_rate.pdf")
plot_locus_amplificatin_rate
dev.off()
```

```{r}
plot_locus_amplificatin_rate
```

Thus in this data set `r ncol(ampseq@gt)` loci had an amplification rate above 0.65, and `r ncol(ampseq_filtered@discarded_loci$gt)` loci were discarded. The discarded loci were: `r paste(colnames(ampseq@discarded_loci$gt), collapse = ', ')`.

# Amplification rate by sample

The proportion of loci amplified per each sample, also called the amplification rate of the samples. The threshold defines the minimum proportion of loci that a sample should have amplified (by default 0.8). FIX THIS TO MAKE IT VARIABLE

```{r}
ampseq_filtered = sample_amplification_rate(ampseq_filtered)

pdf("/cromwell_root/Results/samples_amplification_rate.pdf")
ampseq_filtered@plots$samples_amplification_rate+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12))
dev.off()
```

```{r}
ampseq_filtered@plots$samples_amplification_rate+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12))
```

# plot_log_NPolyLoci_by_pop

WRITE THE TEXT FOR THIS PLOT

```{r}
ampseq_filtered = get_polygenomic(ampseq_object = ampseq_filtered, strata = "Subnational_level2", na.rm = FALSE,
                                  filters = NULL)

plot_log_NPolyLoci_by_pop = log_scale_histogram(data = ampseq_filtered@metadata[ampseq_filtered@metadata$NPolyLoci != 0,],
                                                var = "NPolyLoci", binwidth = 1, group_by = "Subnational_level2",
                                                levels = pop_levels, x_label = "Number of heterozygous loci per sample",
                                                fill_color = pop_colors,
                                                y_breaks = c(1, 5,10, 30), ncol = 4)+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12),
        strip.text = element_text(size = 12)) +
  labs(title = 'Distribution of number of heterozygous loci per sample',
       y = "# Samples",
       x = "# heterozygous loci per sample")

pdf("/cromwell_root/Results/distribution_of_number_heterozygous_loci_per_sample.pdf")
plot_log_NPolyLoci_by_pop
dev.off()
```

```{r}
plot_log_NPolyLoci_by_pop
```

#plot_log_coi_by_pop
WRITE THE TEXT FOR THIS PLOT

```{r}
plot_log_coi_by_pop = log_scale_histogram(data = ampseq_filtered@metadata,
                                          var = "coi",
                                          binwidth = 1,
                                          group_by = "Subnational_level2",
                                          levels = pop_levels, x_label = "COI",
                                          fill_color = pop_colors,
                                          y_breaks = c(1, 10, 100, 400), ncol = 5)+
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 12),
        strip.text = element_text(size = 12)) +
  labs(title = 'Distribution of COI by sampling location',
       y = "# Samples")

pdf("/cromwell_root/Results/distribution_of_COI_by_sampling_location.pdf")
plot_log_coi_by_pop
dev.off()
```

```{r}
plot_log_coi_by_pop
```

Watermark 2
